### ISSUE TITLE
What does input_chunk_length，output_chunk_length and training_length mean respectively?

### ISSUE THREAD

**Issue by Lufffya** [unit8co/darts#1397]
I'm a newbie at darts
These concepts confuse me a lot when I build data and define models.
I am trying training a LSTM Model on Multiple Time Series
![image](https://user-images.githubusercontent.com/49063302/204454863-483299db-6d81-4938-8d4e-976f1e34515e.png)
![image](https://user-images.githubusercontent.com/49063302/204455010-cc722b38-e318-4599-a564-06e75199ffba.png)
In these cases, I can't find the relationship and calculation formula between them, so I don't know how to specify the specific value in my code

**Comment by eliane-maalouf** [comment]
hello, 
training data length is just the total size of the series available for training. From this training series, the models in darts will be expecting a certain size of input from the past (input_chunk_length), this is the length of the window the model can use from the past to predict in the future. Ouput_chunk_length is basically how many steps in the future the model should learn to predict. For example, say you have a series of size 365 days as your training series, you might want your model to learn to use the last 30 days to predict the value of your series 7 days in the future. Here, training length is 365, input_chunk_length = 30 , output_chunk_length = 7. The training series is then cut into multiple input-output pairs of size 30-7 and fed into the algorithm in order for it to learn to reproduce the pattern. Does this clarify those concepts for you?

**Comment by hrzn** [comment]
I want to add that `RNNModel` is somewhat an exception. Contrary to other deep learning models in Darts which work with `input_chunk_length` and `output_chunk_length`, `RNNModel` works with `input_chunk_length` and `training_length`. That's because in a way the output length is always 1 in an RNN (as the sequence model produces outputs one-by-one). For a single backward pass, `training_length` is the length containing the first `input_chunk_length` used to initialise the initial hidden state of the RNN, plus all the following time steps produced by the RNN and on which the loss is computed.

**Comment by Luffffffy** [comment]
@eliane-maalouf
Thank you for you answer！
I probably understand, but this explanation does not seem to apply to all models, for example `RNNModel`, and I have multiple time series with inconsistent length(Not all lengths are 365), and how to specify `training_length`?
(My dataset is more than 1000 time series of three months to three years in length. I want to use `RNNModel`to input 60 days of history to predict the next 30 days)
I guess ` input_chunk_length = 60`，`output_chunk_length = 30`，but `training_length`?

**Comment by Luffffffy** [comment]
> I want to add that `RNNModel` is somewhat an exception. Contrary to other deep learning models in Darts which work with `input_chunk_length` and `output_chunk_length`, `RNNModel` works with `input_chunk_length` and `training_length`. That's because in a way the output length is always 1 in an RNN (as the sequence model produces outputs one-by-one). For a single backward pass, `training_length` is the length containing the first `input_chunk_length` used to initialise the initial hidden state of the RNN, plus all the following time steps produced by the RNN and on which the loss is computed.

This is exactly what confused me. Thank you very much！
I see what you're saying, if so, in `RNNModel`: `output_chunk_length = training_length - input_chunk_length ` Will get the length of the future forecast, right？
But how does the Datrs handle multiple time series of different lengths?

**Comment by hrzn** [comment]
> so, in `RNNModel`: `output_chunk_length = training_length - input_chunk_length ` Will get the length of the future forecast, right？

Kind of but not quite. In `RNNModel` there is no `output_chunk_length` (or it's always 1 if you prefer). The length of the forecast will be determined by the forecast horizon `n` you provide to `predict()`. For training, the loss will be computed on output sequences of length `training_length - input_chunk_length`, so the model is trained to produce sequences of that length, but it can also produce longer sequences auto-regressively. Multiple series are handled by simply slicing sub-sequences of length `training_length` from them and taking the union over all series to get a global training set.

Hope this helps!

**Comment by Luffffffy** [comment]
> Hope this helps!

@hrzn Hello i thought i got it
But when I encountered the following error, I doubted again

`The dataset contains some time series that are too short to contain max(self.input_chunk_length, self.shift + self.output_chunk_length)` 

My requirement is to output 60 days, make it output 30 days and calculate the loss on the output sequence (I know that RNN can loop and output predictions of any length) so my relevant parameters are

```
        model = RNNModel(
            model='LSTM',
            input_chunk_length=input_chunk_length,
            hidden_dim=hidden_dim,
            n_rnn_layers=n_rnn_layers,
            dropout=dropout,
            training_length=input_chunk_length + val_len - 1,
            n_epochs=max_n_epochs,
            batch_size=batch_size,
            loss_fn=torch.nn.MSELoss(),
            random_state=seed,
            optimizer_kwargs={'lr': lr},
            model_name=model_name,
            force_reset=True,
            save_checkpoints=True,
            pl_trainer_kwargs=pl_trainer_kwargs
        )

 
        model.fit(
                    series=train,
                    val_series=model_val_set,
                    future_covariates=training_covariates,
                    val_future_covariates=training_covariates
         )
```

The most important of these parameters are  `input_chunk_length = 60`  `val_len = 30`
And the shortest time series length for training is 150

Take a time series as an example：
The length of the split training test set is 
`train = [0: 90]`  `model_val_set =[ 30 :120 ]` and Reserve the verification length of the last 30

I noticed the self.shift parameter, which might be why I'm failing, but I don't know what it's specifically worth

**Comment by madtoinou** [comment]
Hi @Lufffya,

I could not reproduce your error with the latest darts version with the following code snippet:

```python
from darts.models import RNNModel
from darts.datasets import AirPassengersDataset
import darts.utils.timeseries_generation as tg

series = AirPassengersDataset().load()
input_chunk_length = 60
val_len = 30

train = series[:90]
val = series[30:120]
covs = tg.gaussian_timeseries(length=120, start=series.start_time(), freq=series.freq)

model = RNNModel(
            model='LSTM',
            input_chunk_length=input_chunk_length,
            training_length=input_chunk_length + val_len - 1,
            n_epochs=1,
            loss_fn=torch.nn.MSELoss(),
            force_reset=True,
        )

 
model.fit(
            series=train,
            val_series=val,
            future_covariates=covs,
            val_future_covariates=covs
    )
```

Closing this for now, feel free to reopen it or open a new issue if you have additional questions.
